pkg/jmespath/jmespath.go

```diff
@@ -0,0 +1,201 @@
+	if e.files != nil {
+		// Don't mutate the original map
+		copy := make(map[string]any, len(data)+1)
+		maps.Copy(copy, data)
+		copy[filesKey] = e.files.getMetadata()
+		data = copy
+	}
+	return e.expression.Search(data)
+}
+
+func (e *Expression) checkTestVector(t *pb.TestVector) error {
+	input := t.Input.AsMap()
+
+	files, filesPresent := input[filesKey].(map[string]any)
+	if e.files == nil {
+		if filesPresent {
+			return status.Error(codes.InvalidArgument, "Test vector contains file contents, but no files were provided in the JMESPath expression configuration")
+		}
+	} else {
+		if !filesPresent {
+			return status.Errorf(codes.InvalidArgument, "Test vector input is missing %q key", filesKey)
+		}
+		for _, key := range e.files.files {
+			if _, ok := files[key.Key]; !ok {
+				return status.Errorf(codes.InvalidArgument, "Test vector is missing file %q", key.Key)
+			}
+		}
+		if len(files) != len(e.files.files) {
+			return status.Errorf(codes.InvalidArgument, "Test vector contains %d files, but JMESPath expression expects %d files", len(files), len(e.files.files))
+		}
+	}
+
+	actual, err := e.expression.Search(input)
+	if err != nil {
+		return util.StatusWrapf(err, "Failed to evaluate JMESPath expression on test vector input")
+	}
+
+	actualPb, err := structpb.NewValue(actual)
+	if err != nil {
+		return util.StatusWrapf(err, "Failed to convert JMESPath result to protobuf value")
+	}
+
+	if !reflect.DeepEqual(t.ExpectedOutput, actualPb) {
+		expectedJSON, _ := json.Marshal(t.ExpectedOutput)
+		actualJSON, _ := json.Marshal(actual)
+		return status.Errorf(codes.InvalidArgument, "Test vector failed: expected %s, got %s", string(expectedJSON), string(actualJSON))
+	}
+
+	return nil
+}
+
+// metadataFileProvider makes the contents of files available as
+// metadata to JMESPath expressions.
+type metadataFileProvider struct {
+	lock            sync.RWMutex
+	files           []*pb.File
+	currentContents map[string]string
+}
+
+// newJMESPathMetadataFileProvider creates a MetadataFileProvider that
+// reads files from the filesystem and makes their contents available
+// as metadata. The contents of the files are reloaded periodically.
+func newJMESPathMetadataFileProvider(files []*pb.File, group program.Group) (*metadataFileProvider, error) {
+	provider := &metadataFileProvider{
+		files:           files,
+		currentContents: make(map[string]string),
+	}
+	for _, file := range files {
+		contents, err := readFile(file.Path)
+		if err != nil {
+			return nil, err
+		}
+		provider.currentContents[file.Key] = contents
+		refresher := func(ctx context.Context) error {
+			t := time.NewTicker(60 * time.Second)
+			defer t.Stop()
+			for {
+				select {
+				case <-t.C:
+					contents, err := readFile(file.Path)
+					if err != nil {
+						log.Printf("Failed to reload %s file: %v", file.Path, err)
+					} else {
+						provider.lock.Lock()
+						provider.currentContents[file.Key] = contents
+						provider.lock.Unlock()
```

Comment:
What are your thoughts on spawning a single coroutine per expression, as opposed to one per file? Then that goroutine could construct the entirely new "files" map, as opposed modifying the existing map in place. The advantage is that you can then use a `sync.Atomic[map[string]any]` instead of `sync.RWMutex`.

Reply:
Sounds like a good idea - I've done this.